### PR TITLE
Support for unused model initializers

### PR DESCRIPTION
**Pull request by baijumeswani**
**Description**: If a model has unused ```named_parameters```, the initializer is dropped after model export. But the ```ORTModule``` frontend communicates all PyTorch model initializers to the backend. The backend is unable to find the initializers that were dropped and this results in an error message and an exception is thrown.
To resolve this problem through this pull request, the frontend needs to pass only the initializers that are recognized in the onnx graph to the backend.

### CODE REVIEW

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 1]
```diff
@@ -2252,3 +2252,50 @@ def run_step(expected, a, b, c, d, e, f, y, z):
...
+    ort_model.eval()
+
+    x = torch.randn(N, D_in, device=device)
+    y = copy.deepcopy(x)
+
+    # Make sure model runs without any exception
+    out_pt = model(x)
+    out_ort = ort_model(y)
+    assert out_ort is not None
+    assert torch.equal(out_ort, out_pt)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:2299]: we cannot use torch.equal compare float numbers... Is 0 == 0.0000000001 ? 
please use _test_helpers.assert_values_are_close() instead.... (this use torch.allclose under the hood)

Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:2299]: sure, I can update to ```_test_helpers.assert_values_are_close()``` instead.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 2]
```diff
@@ -2252,3 +2252,50 @@ def run_step(expected, a, b, c, d, e, f, y, z):
...
+    ort_model = ORTModule(copy.deepcopy(model))
+
+    x = torch.randn(N, D_in, device=device)
+    y = copy.deepcopy(x)
+
+    # Make sure model runs without any exception
+    out_pt = model(x)
+    out_ort = ort_model(y)
+    assert out_ort is not None
+    assert torch.equal(out_ort, out_pt)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:2284]: Same here... 
please replace all the torch.equal with _test_helpers.assert_values_are_close for output/gradient/module_weight comparision. 

unless we one tensor is a copy of another tensor, we shouldn't use torch.equal for comparison.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 3]
```diff
@@ -2252,3 +2252,50 @@ def run_step(expected, a, b, c, d, e, f, y, z):
...
+    out_ort.backward()
+
+    # Also try in eval mode
+    model.eval()
+    ort_model.eval()
+
+    x = torch.randn(N, D_in, device=device)
+    y = copy.deepcopy(x)
+
+    # Make sure model runs without any exception
```
Comment (SherlockNoMad) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:2295]: please also do a few training steps and compare it's gradient value for correctness.

**Review by SherlockNoMad — CHANGES_REQUESTED** [review]
update UT please

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 4]
```diff
@@ -267,9 +267,14 @@ def _initialize_graph_builder(self, training):
...
-        initializer_names_to_train = [name for name,
-                                      param in self._flattened_module.named_parameters() if param.requires_grad]
+        # All initializer names along with user inputs are a part of the onnx graph inputs
+        # since the onnx model was exported with the flag keep_initializers_as_inputs=True
+        onnx_initializer_names = {p.name for p in self._onnx_model.graph.input}
+
+        initializer_names = [name for name, _ in self._flattened_module.named_parameters()
+                             if name in onnx_initializer_names]
+        initializer_names_to_train = [name for name, param in self._flattened_module.named_parameters()
+                                      if param.requires_grad and name in onnx_initializer_names]
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:277]: I am not sure if printing a warning here is a good idea. Consider an example:
```py
def forward(self, bool_input, input1):
    if bool_input:
        do_something_with_fc1
    else:
        do_somethind_else_with_fc2
```

For this example, whether the user enters ```bool_input``` as ```True``` or ```False```, outputting a warning might be informative, but also unnecessary. In what scenario do you think it would be detrimental to the user with a warning not printed?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 5]
```diff
@@ -2362,3 +2362,51 @@ def forward(self, bool_argument, input1):
 
     # Ensure that no exceptions are raised
     out = model(bool_argument, x)
+
+def test_unused_parameters():
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:2366]: Actually, we combine inputs, buffers and then initializers (in that order). Having said that, we can definitely add a buffer to this test to make sure that the inputs are combined without any trouble.

[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:2366]: My bad lol I was sure it was the other way around
